File: javascript/phone-number/phone-number.js
const MAX_NUMBER = 11;
const LENGTH = 10;
const AREA_INDEX = 0;
const EXCHANGE_INDEX = 3;

export const clean = input => {
    let output = '';
    if (/[\:$%\^@!]/.test(input)) throw new Error('Punctuations not permitted');
    if (/[a-z]/i.test(input)) throw new Error("Letters not permitted"); 
    for (const [index, value] of [...input.replace(/^1|^\+1|[()-\s\.]/g,'')].entries()){
        const digit = parseInt(value);
        if ((index == AREA_INDEX || index == EXCHANGE_INDEX) && digit < 2) 
          throw new Error(`${index == AREA_INDEX ? 'Area': 'Exchange'} code cannot \
start with ${digit == AREA_INDEX ? 'zero' : 'one'}`);
        output += digit;
      }
    if (output.length === MAX_NUMBER) throw new Error(`${MAX_NUMBER} digits must start with 1`);
    if (output.length > MAX_NUMBER) throw new Error(`More than ${MAX_NUMBER} digits`);
    if (output.length !== LENGTH) throw new Error('Incorrect number of digits');

    return output;
}

File: python/phone-number/phone_number.py
import re


class PhoneNumber:
    def __init__(self, number):
        self.number = re.sub(r'^1|(^\+1)|[()\-\s\.]', '', number)

        if len(self.number) < 10:
            raise ValueError("must not be fewer than 10 digits")
        elif len(self.number) > 11:
            raise ValueError("must not be greater than 11 digits")
        elif len(self.number) == 11:
            raise ValueError("11 digits must start with 1")
        elif re.search(r'[^A-Za-z0-9]', self.number):
            raise ValueError("punctuations not permitted")
        elif re.search(r'[^\d]', self.number):
            raise ValueError("letters not permitted")
        elif int(self.number[0]) < 2:
            raise ValueError(f"area code cannot start with \
{'one' if int(self.number[0]) else 'zero'}")
        elif int(self.number[3]) < 2:
            raise ValueError(f"exchange code cannot start with \
{'one' if int(self.number[3]) else 'zero'}")
        self.area_code = self.number[0:3]

    def pretty(self):
        return f'({self.area_code})-{self.number[3:6]}-{self.number[6:10]}'
